design/user_controller_workspaces.md

```diff
@@ -0,0 +1,143 @@
+
+It is the user's responsibility to ensure they commit at the end of their job/pipeline/script,
+but it's possible to pass around that `workspaceID` between jobs and processes, even if reading/writing concurrently.
+
+Once a commit is done for a given workspace, it should no longer be used. if using the lakeFS SDK (as shown above),
+calling any operation on a `commit()`ed `Workspace` object will raise an error. We can go further and place a "lockfile" (`_lakefs_committed/${workspaceID}`) and check it before writing if we think this is neccesary (on a best effort basis).
+
+## Using the S3 gateway
+
+just append a workspace ID: `s3://my-repo/master+abc123/my/collection/` to read/write.
+otherwise, just pass a ref as you normally would, to read committed data.
+
+## Out-of-band ingest
+
+Now that a commit is basically a diffIterator on top of `listObjects` we can do something similar for stream processing.
+Working with streams is different: we can't always ensure no writes are happening, but fortunately, most streaming processors 
+do [guarantee immutability](https://stackoverflow.com/a/61201083): A written file is never overwritten, only written once and potentially deleted: this allows us to take
+a regular old object store path and do:
+
+```sh
+$ lakectl ingest commit \
+    --from-path "s3://bucket/kafka/events/2021/" \
+    --to-path "lakefs://my-repo@ingest-branch/kafka/events/2021/" \
+    -m "ingest events: $(date -u)"
+
+# Or just see what changes would be introduced
+$ lakectl ingest diff \
+    --from-path "s3://bucket/kafka/events/2021/" \
+    --to-path "lakefs://my-repo@ingest-branch/kafka/events/2021/" \
+```
+Here, instead of treating the given object store location as a diffIterator, we look at it as a subtree, to be diffed with the underlying commit.
+
+
+## IAM permissions
+
+set them normally on `s3://storagenamespace/*/path/a`, `s3://storagenamespace/*/path/b`, etc.
+
+
+## S3 Lifecycle rules
+
+when writing, we can pass tags using the SDK. let's say we write to `ws.path("foo/bar/baz.parquet")`:
+we can tag that write with `path[0]=foo,path[1]=bar,path[2]=baz.parquet`
+
+Now we can set a lifecycle rule: delete after 30 days if tags match:
+
+```xml
+<Tags>
+    <Tag>
+        <key>path[0]</key>
+        <value>foo</key>
+    </Tag>
+    <Tag>
+        <key>path[1]</key>
+        <value>bar</key>
+    </Tag>
+</Tags>
+```
+
+We can provide tooling to generate or configure those on behalf of the user.
+
+## Standalone mode
+
+Running on a single instance we don't *need* a CAS operation, we can serialize access to the ref keys in-process, so a ref could sit in any object store or local directory. `lakefs -dev` now works without a database (I'd still show a big red warning in the UI when running like this - also true for the local:// adapter in general).
+
+
+## Uploading/diff staged in the UI
+
+When uploading objects we can generate a workspace ID and stick it in `localStorage`.
```

Comment:
However there is no list of "current staging" (`git stash list`), so hard to use the UI to finagle about with existing workspaces.  This will make debugging broken pipelines harder.